crates/runtime/src/externref.rs

```diff
@@ -393,6 +461,559 @@ impl Deref for VMExternRef {
+            //
+            // This overwrites old entries, which drops them and decrements their
+            // reference counts. Safety relies on the reference count increment in
+            // `insert_precise_stack_root` to avoid over-eagerly dropping references
+            // that are in `self.precise_stack_roots` but haven't been inserted into
+            // the bump allocation table yet.
+            let mut precise_roots = precise_roots.drain();
+            'outer: for (chunk_index, chunk) in chunks.iter().enumerate() {
+                for (slot_index, slot) in chunk.iter().enumerate() {
+                    if let Some(root) = precise_roots.next() {
+                        unsafe {
+                            // NB: there is no reference count increment here
+                            // because everything in `self.precise_stack_roots`
+                            // already had its reference count incremented for us,
+                            // and this is logically a move out from there, rather
+                            // than a clone.
+                            *slot.get() = Some(VMExternRef(root));
+                        }
+                    } else {
+                        // We've inserted all of our precise, on-stack roots back
+                        // into the bump allocation table. Update our `next` and
+                        // `end` bump pointer members for the new current chunk, and
+                        // free any excess chunks.
+                        let start = chunk.as_ptr() as *mut TableElem;
+                        unsafe {
+                            let next = start.add(slot_index + 1);
+                            debug_assert!(!next.is_null());
+                            *self.next.get() = NonNull::new_unchecked(next);
+
+                            let end = start.add(chunk.len());
+                            debug_assert!(!end.is_null());
+                            self.end.set(NonNull::new_unchecked(end));
+                        }
+                        chunks.truncate(chunk_index + 1);
+                        break 'outer;
+                    }
+                }
+            }
+
+            debug_assert!(
+                precise_roots.next().is_none(),
+                "should always have enough capacity in the bump allocations table \
+                 to hold all of our precise, on-stack roots"
+            );
+        }
+
+        // Finally, sweep away excess capacity within our new last/current
+        // chunk, so that old, no-longer-live roots get dropped.
+        let num_filled_in_last_chunk = self.num_filled_in_last_chunk(&chunks);
+        for slot in chunks.last().unwrap().iter().skip(num_filled_in_last_chunk) {
+            unsafe {
+                *slot.get() = None;
+            }
+        }
+    }
+}
+
+/// A registry of stack maps for currently active Wasm modules.
+#[derive(Default)]
+pub struct StackMapRegistry {
```

Comment:
We've got a lot of various registries floating around wasmtime, would it be possible to insert this into an existing registry somehow? For example most of the code here looks like it's copying what we do for frame information.